File: Tool.java
/**
 * @author Ben Irastorza
 * @version 1.1337
 */
/*
    * The program does NOT require you to surrender your instagram username and password.
    * The program also is all ran locally and securely on your personal computer.
    * The user is only required to input through the command line.

    * Step 1: Download your Instagram data per the instructions on https://help.instagram.com/181231772500920
    *  MAKE SURE TO CHOOSE "JSON" FILE FORMAT AND SELECT "ALL TIME" DATA
    * Step 2: Get the path of the Following and Followers json files under "followers_and_following"
    * Step 3: Create an empty .txt file and copy the file path
    * Step 4: Run the program and paste the 3 required file paths when requested
 */
import java.io.File;
import java.io.FileNotFoundException;
import java.util.HashSet;
import java.util.Scanner;
import java.util.ArrayList;
import java.io.PrintWriter;

public class Tool {
    /**
     * @param args main method parameter
     */
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String followingPath;
        String followersPath;
        String txt;

        try {
            System.out.print("Input path to following JSON file: ");
            followingPath = sc.nextLine();
            System.out.print("Input path to followers JSON file: ");
            followersPath = sc.nextLine();
            System.out.print("Input path to empty .txt file: ");
            txt = sc.nextLine();
            System.out.println("");

            ArrayList<String> followers = followersToArr(followersPath);
            ArrayList<String> following = followingToArr(followingPath);
            ArrayList<String> answer = notFollowBack(following, followers);
            writeToFile(new File(txt), answer);

        } catch (RuntimeException re) {
            System.out.println("One of the file paths was incorrect");
        }
    }

    /**
     * followingToArr parses the JSON file and creates a list of the users following.
     * @param filePath the path to the following file
     * @return the list of following
     */
    public static ArrayList<String> followingToArr(String filePath) {
        ArrayList<String> toReturn = new ArrayList<>();
        File x = new File(filePath);
        try {
            String currLine = "";
            Scanner scan = new Scanner(x);
            while (scan.hasNextLine()) {
                currLine = scan.nextLine();
                if (currLine.contains("value")) {
                    currLine = currLine.substring(20, currLine.length() - 2);
                    toReturn.add(currLine);
                }
            }
            return toReturn;
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * followersToArr parses the JSON file and creates a list of the users followers.
     * @param filePath the path to the followers file
     * @return the list of followers
     */
    public static ArrayList<String> followersToArr(String filePath) {
        ArrayList<String> toReturn = new ArrayList<>();
        File x = new File(filePath);
        try {
            String currLine = "";
            Scanner scan = new Scanner(x);
            while (scan.hasNextLine()) {
                currLine = scan.nextLine();
                if (currLine.contains("value")) {
                    currLine = currLine.substring(18, currLine.length() - 2);
                    toReturn.add(currLine);
                }
            }
            return toReturn;
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * The following method loops through both lists adding the flagged usernames to a new list.
     * Optimization: updated runtime to O(n)
     * @param following the list of following
     * @param followers the list of followers
     * @return the list of people not following the user back
     */
    public static ArrayList<String> notFollowBack(ArrayList<String> following, ArrayList<String> followers) {
        ArrayList<String> answer = new ArrayList<>();
        HashSet<String> followersSet = new HashSet<>(followers);
        for (String s : following) {
            if (!followersSet.contains(s)) {
                answer.add(s);
            }
        }
        return answer;
    }

    /**
     * writeToFile writes the data to a text file for the user to have.
     * @param x the file the user chose to have the data written to.
     * @param answer the list of the people that don't follow the user back.
     */
    public static void writeToFile(File x, ArrayList<String> answer) {
        try {
            PrintWriter pw = new PrintWriter(x);
            for (String s : answer) {
                pw.println(s);
            }
            pw.close();
            System.out.println("The data has successfully been written to the text file.");
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
    }
}

File: Tool.py
import json
import os

def followers_to_set(followers_path):
    """
    Read followers data from a JSON file and return a set of follower usernames.

    Parameters:
    followers_path (str): Path to the JSON file containing followers data.

    Returns:
    set: A set of follower usernames.
    """
    try:
        with open(followers_path) as f:
            data = json.load(f)
        values = set()
        for item in data:
            value = item['string_list_data'][0]['value']
            values.add(value)
        return values
    except Exception as e:
        print(f"Error reading followers data from '{followers_path}': {e}")

def following_to_list(following_path):
    """
    Read following data from a JSON file and return a list of following usernames.

    Parameters:
    following_path (str): Path to the JSON file containing following data.

    Returns:
    list: A list of following usernames.
    """
    try:
        with open(following_path) as file:
            data = json.load(file)
        values = []
        for item in data['relationships_following']:
            value = item['string_list_data'][0]['value']
            values.append(value)
        return values
    except Exception as e:
        print(f"Error reading following data from '{following_path}': {e}")

def write_to_text(followers, following):
    """
    Write the usernames of users not following back to a text file.

    Parameters:
    followers (set): Set of follower usernames.
    following (list): List of following usernames.
    """
    not_following_back = []
    for user in following:
        if user not in followers:
            not_following_back.append(user)
    file_path = os.path.join(os.path.dirname(__file__), "not_following_back.txt")
    with open(file_path, 'w') as file:
        for item in not_following_back:
            file.write(item + '\n')
    print("Success: Results written to 'not_following_back.txt'")
    print("The file is located at: " + os.path.dirname(os.path.abspath(__file__)))

# User Input
followers_path = input("Input path to followers JSON file: ")
following_path = input("Input path to following JSON file: ")
print()

# Main Execution
try:
    followers_set = followers_to_set(followers_path)
    following_list = following_to_list(following_path)
    write_to_text(followers_set, following_list)
except Exception as e:
    print(f"Failure: {e}")
